internal/engine/actions/actions.go

```diff
@@ -138,22 +138,82 @@ func (rae *RuleActionsEngine) processAction(
+		switch actionType {
+		case remediate.ActionType:
+			// Return the previous remediation status.
+			err := enginerr.RemediationStatusAsError(params.GetEvalStatusFromDb().RemStatus.RemediationStatusTypes)
+			return *(getMeta(params.GetEvalStatusFromDb().RemMetadata)), err
+		case alert.ActionType:
+			// Return the previous alert status.
+			err := enginerr.AlertStatusAsError(params.GetEvalStatusFromDb().AlertStatus.AlertStatusTypes)
+			return *(getMeta(params.GetEvalStatusFromDb().AlertMetadata)), err
+		}
+	}
 	// Get action engine
 	action := rae.actions[actionType]
 	// Return the result of the action
-	logger := zerolog.Ctx(ctx)
-	logger.Debug().
-		Str("action", string(actionType)).
-		Str("cmd", string(cmd)).
-		Msg("invoking action")
+	logger.Msg("invoking action")
 	return action.Do(ctx, cmd, rae.actionsOnOff[actionType], ent, params, metadata)
 }
 
 // shouldRemediate returns the action command for remediation taking into account previous evaluations
 func shouldRemediate(prevEvalFromDb *db.ListRuleEvaluationsByProfileIdRow, evalErr error) engif.ActionCmd {
-	_ = prevEvalFromDb
-	_ = evalErr
-	return engif.ActionCmdOn
+	// Get current evaluation status
+	newEval := enginerr.ErrorAsEvalStatus(evalErr)
+
+	// Get previous evaluation status
+	prevEval := db.EvalStatusTypesPending
+	if prevEvalFromDb.EvalStatus.Valid {
+		prevEval = prevEvalFromDb.EvalStatus.EvalStatusTypes
+	}
+
+	// Get previous Remediation status
+	prevRemediation := db.RemediationStatusTypesSkipped
+	if prevEvalFromDb.AlertStatus.Valid {
+		prevRemediation = prevEvalFromDb.RemStatus.RemediationStatusTypes
+	}
+
+	// Start evaluation scenarios
+
+	// Case 1 - Do nothing if the evaluation status has not changed and remediation did not errored-out
+	if newEval == prevEval && prevRemediation != db.RemediationStatusTypesError {
+		return engif.ActionCmdDoNothing
+	}
+
+	// Proceed with use cases where the evaluation changed
+	switch newEval {
+	case db.EvalStatusTypesError:
+	case db.EvalStatusTypesSuccess:
+		// Case 2 - Evaluation changed from something else to ERROR -> Remediation should be OFF
+		// Case 3 - Evaluation changed from something else to PASSING -> Remediation should be OFF
+		// The Remediation should be OFF (if it wasn't already)
+		if db.RemediationStatusTypesSkipped != prevRemediation {
+			return engif.ActionCmdOff
+		}
+		// We should do nothing if remediation was already skipped
+		return engif.ActionCmdDoNothing
+	case db.EvalStatusTypesFailure:
+		// Case 4 - Evaluation has changed from something else to FAILED -> Remediation should be ON
+		// We should do nothing if the Remediation is already pending or successful
+		if db.RemediationStatusTypesPending == prevRemediation || db.RemediationStatusTypesSuccess == prevRemediation {
+			return engif.ActionCmdDoNothing
+		}
+		// The Remediation should be turned ON (if it wasn't already)
+		return engif.ActionCmdOn
+	case db.EvalStatusTypesSkipped:
+	case db.EvalStatusTypesPending:
```

Comment:
can we even get pending eval here (its fine to leave it there, just asking)

Reply:
I don't think so too, but the lint check said it's a good practice so thus why I added it 👍